tests/basic.test.tsx

```diff
@@ -775,10 +775,48 @@ it('should have correct (partial) types for setState', () => {
   // ok, should not error
   store.setState({ count: 1 })
   store.setState({})
-  store.setState(() => {})
+  store.setState(() => undefined)
+  store.setState((previous) => previous)
 
   // @ts-expect-error type undefined is not assignable to type number
   store.setState({ count: undefined })
   // @ts-expect-error type undefined is not assignable to type number
   store.setState((state) => ({ ...state, count: undefined }))
 })
+
+it('should allow for different partial keys to be returnable from setState', () => {
+  type State = { count: number; name: string }
+
+  const store = create<State>(() => ({
+    count: 0,
+    name: 'foo',
+  }))
+
+  const setState: AssertEqual<typeof store.setState, SetState<State>> = true
+  expect(setState).toEqual(true)
+
+  // ok, should not error
+  store.setState((previous) => {
+    if (previous.count === 0) {
+      return { count: 1 }
+    }
+    return { count: 0 }
+  })
+  store.setState<'count', 'name'>((previous) => {
+    if (previous.count === 0) {
+      return { count: 1 }
+    }
+    if (previous.count === 1) {
+      return previous
+    }
+    return { name: 'foo' }
+  })
+
+  // @ts-expect-error type  Type '{ name: boolean; count?: undefined; }' is not assignable to type State.
+  store.setState<'count', 'name'>((previous) => {
+    if (previous.count === 0) {
+      return { count: 1 }
+    }
+    return { name: true }
+  })
+})
```

Comment:
i would expect an error here, because `name: true` certainly does not fit. However, there is no error, and I don't really know why. it errors fine the [ts playground](https://www.typescriptlang.org/play?ts=4.2.3#code/KYDwDg9gTgLgBDAnmYcDKMCGNUF44QBGAVsAMYwBQA9NQssAM5wCWAdo2C1MACZyFEcRsBgZsqAGZQIAWzgAlYJgoAaYcGAAuGnQAWMGGEZbaAcxYw9AV0IA6MnOoARYJPaXgAG0QAVBrwubh44Pv4ogYReRNQAbGQALACcwAkArAAcCQBMAMyZhJhpSUUADBnAAIy51aUA7EmVSRlJjZUJuaWBSCiM1DwqMNQAbpWx1Oy8oHa8djCMAMQAMgktALQrSWmUoJCw9ChwAAqYsCyYXuI4ADy+cKA4bLzMV8DqANKV9yCPz3AA1sBEBBJHA7vhAcDQb4Ptlvr9mJCQWC4PhPh9cvDgE9EUDkeC4O9sh8ElicQC8dDUYTcgA+VGUOBwAA+cAAFEcWGR-rcPpV6azOdzeYTsgLjlyeTCaeKhVKSeLfABKRks9lsxhYHBaMFK1H0uUiz6yyVGsVqw3S950i2mq0JRUq3bQeA9VCvNDecgwaC3Ml-V7qACq9PwGq12l1+rgQZ24BdB1QAFEAI7WC6WRAAYT05EBUFuofZmokOulbGAAHdXmW9bh6YQIBAvMo2HG9q6GOgI0sWJrscAC74i+HS2D1GAeMMWBBrIwa1H63BhhAWLx2wm3d2JGgvFzgL3+xWhyPGHuyJHpZPgNPZ-Pz5e6-SV2uN-t2DgoJIVO7bIwyFALCEMAfoPNiAYRvSADeqpsnuR6DjqryHo8g6FkqOpsk+y6ruuTLXB6D60myqpMvBqFQEhEa7vuKEDgWhH7rSqikRoLYUNAAD8VE7l6HFDuojEXsxrHAGmGZIAAYmw3FwKm6Z7kgOZ5mhQnALSqoYey2EvuuAC+lBvp2hyemIEagT84EvJB1Iwfhnz+riUIohClLjqKjkUs5BLojSnlIlSaLEoSpJgeSAUuTKbJgKcMDnF4OonGcFyvEalSwhiCrqDwYBeD+smNs2rZabplAGc6+xbgA4qIqV3GFEESCO2G+EZiZwK4moyEIYY6bhbUfoO34Xt20DAAAglwFkItuODQaqIhmWOpl1RpTJmLVEY6jVS03MOC1-gBQGRmgh2AcBhaqlMXUQIgOqdTA3VlW1W6vFmAw+kOnmBnAWZzj6sgrRG1JAxIhYjgtog6n9mpyKDOAsetUNwDtq2I3AmBcFRY2TSwl0tYZlCOBw8Cwzw1IAES8NYsiyIgFMY8w1hsP8bAQJWbCM6NPC49cUFwJITY6l17BmAA3AIpw6mwNPAVAcB6RpZPAHYi2pQA5ILEDq+o6uFFA6vEdFU4znO2F2awoLGzepuMHY+uorg+CVHqFtMjwMDWFAnP81rOoU1rDN6WLqoGe7ohez7kuUQgUDWKgwdlUqIeUEAA)

@dai-shi do you see why?

Reply:
ahm okay, found it. seems to because of `"name"` ? If I rename the prop to something different, it errors correctly :/

cd1a1fa7

Reply:
so, "name" is something special in ts??

Reply:
no, not that I know of. Have no idea tbh

Reply:
FYI
"name" is a global: https://developer.mozilla.org/en-US/docs/Web/API/Window/name
makes for many annoying false-positives when using TypeScript 😄